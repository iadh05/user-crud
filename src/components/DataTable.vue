<template>
    <div>
        <v-row>
            <v-col cols="3">
                <v-text-field label="search" v-model="search"></v-text-field>
            </v-col>
        </v-row>

        <v-data-table
            class="user-table"
            :headers="headers"
            :items="items"
            item-key="name"
            no-data-text="No data"
            no-results-text="No user found"
            :search="search"
            :loading="isLoading"
            @pagination="pagination"
        >
            <template v-slot:item="{item}">
                <tr>
                    <td data-cy="id">
                        <TextHighlight :text="item.id" :search="search" />
                    </td>
                    <td data-cy="firstName">
                        <TextHighlight
                            :text="item.firstName"
                            :search="search"
                        />
                    </td>
                    <td data-cy="lastName">
                        <TextHighlight :text="item.lastName" :search="search" />
                    </td>
                    <td data-cy="email">
                        <TextHighlight :text="item.email" :search="search" />
                    </td>
                    <td data-cy="address">
                        <TextHighlight :text="item.address" :search="search" />
                    </td>
                    <td class="actions-container">
                        <v-icon
                            class="table-actions"
                            @click="deleteUser(item.id)"
                            >mdi-delete</v-icon
                        >
                        <UserForm
                            :icon="true"
                            :userId="item.id"
                            :update="true"
                        />
                        <UserForm
                            :icon="true"
                            :userId="item.id"
                            :readonly="true"
                        />
                    </td>
                </tr>
            </template>
        </v-data-table>
    </div>
</template>

<script>
import UserForm from "./UserForm.vue";
import TextHighlight from "./TextHighlight.vue";
export default {
    name: "Table",
    components: {
        TextHighlight,
        UserForm,
    },
    props: {
        items: Array,
        headers: Array,
        isLoading: Boolean,
    },
    data() {
        return {
            search: null,
        };
    },
    methods: {
        pagination(p) {
            this.length = p.itemsLength;
        },
        deleteUser(id) {
            this.$emit("deleteByUserId", id);
        },
    },
};
</script>

<style>
.user-table .table-actions {
    cursor: pointer;
}
.user-table .actions-container {
    display: flex;
    align-items: center;
}
.user-table.theme--light.v-data-table
    > .v-data-table__wrapper
    > table
    > thead
    > tr
    > th {
    font-size: 11px;
    color: rgb(0 0 0);
    height: 30px;
    padding: 0 3px;
    text-align: left;
}
td {
    text-align: left;
}
.v-application--is-ltr .user-table .v-data-footer__select {
    margin-left: 0;
    margin-right: auto;
}
</style>
