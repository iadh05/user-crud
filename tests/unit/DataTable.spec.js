import Vue from 'vue'
import { mount, createLocalVue, shallowMount } from '@vue/test-utils'
import DataTable from '@/components/DataTable.vue'
import Vuetify from 'vuetify'
import Vuex from "vuex"

const localVue = createLocalVue()
const users = [
    {
        id: 1,
        firstName: "Joe",
        lastName: 'Doe',
        email: "test@email.com",
        address: "test"
    },
    {
        id: 2,
        firstName: "iadh",
        lastName: 'ben abderrahman',
        email: "iadh@email.com",
        address: "30 sproles"
    },
]
Vue.use(Vuetify);
let actions
let store
let getters;
let state
const headers = [{ "text": "id", "value": "id" }, { "text": "firstName", "value": "firstName" }, { "text": "lastName", "value": "lastName" }, { "text": "email", "value": "email" }, { "text": "address", "value": "address" }, { "text": "actions", "value": "" }]
Vue.use(Vuetify);
describe('DataTable.vue', () => {
    beforeEach(() => {
        localVue.use(Vuex)
        actions = {
            updateUserFields: jest.fn(),
            updateUser: jest.fn()
        }
        getters = {
            getUser: (state) => ({ field }) => {
                return state.user[field];
            },

        }
        state = {
            user: {
                firstName: "jhon",
                lastName: "doe",
                email: "test@gmail.com",
                address: '30 street'
            }
        }

        store = new Vuex.Store({
            actions,
            getters,
            state
        })


    })
    it('should show Table', () => {
        const wrapper = mount(DataTable, {
            store,
            localVue,
            vuetify: new Vuetify,
            propsData: {
                items: users,
                headers
            }

        })

        expect(wrapper.html()).toMatchSnapshot()
    })
    it('should show Table result', () => {
        const wrapper = mount(DataTable, {
            store,
            localVue,
            vuetify: new Vuetify,
            propsData: {
                items: users,
                headers
            },
            data() {
                return {
                    search: "iadh"
                }
            }

        })

        expect(wrapper.find("[data-cy='firstName']").text()).toMatch(wrapper.vm.search);
    })

})

