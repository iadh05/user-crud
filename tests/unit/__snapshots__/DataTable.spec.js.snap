// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTable.vue should show Table 1`] = `
<div>
  <div class="row">
    <div class="col col-3">
      <div class="v-input theme--light v-text-field">
        <div class="v-input__control">
          <div class="v-input__slot">
            <div class="v-text-field__slot"><label for="input-4" class="v-label theme--light" style="left: 0px; position: absolute;">search</label><input id="input-4" type="text"></div>
          </div>
          <div class="v-text-field__details">
            <div class="v-messages theme--light">
              <transition-group-stub tag="div" name="message-transition" class="v-messages__wrapper"></transition-group-stub>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div class="v-data-table user-table v-data-table--has-bottom theme--light">
    <div class="v-data-table__wrapper">
      <table>
        <colgroup>
          <col class="">
          <col class="">
          <col class="">
          <col class="">
          <col class="">
          <col class="">
        </colgroup>
        <thead class="v-data-table-header">
          <tr>
            <th role="columnheader" scope="col" aria-label="id: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>id</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="firstName: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>firstName</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="lastName: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>lastName</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="email: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>email</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="address: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>address</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
            <th role="columnheader" scope="col" aria-label="actions: Not sorted. Activate to sort ascending." aria-sort="none" class="text-start sortable"><span>actions</span><i aria-hidden="true" class="v-icon notranslate v-data-table-header__icon mdi mdi-arrow-up theme--light" style="font-size: 18px;"></i></th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td data-cy="id"><span class="text-highlight">1</span></td>
            <td data-cy="firstName"><span class="text-highlight">Joe</span></td>
            <td data-cy="lastName"><span class="text-highlight">Doe</span></td>
            <td data-cy="email"><span class="text-highlight">test@email.com</span></td>
            <td data-cy="address"><span class="text-highlight">test</span></td>
            <td class="actions-container"><button type="button" class="v-icon notranslate table-actions v-icon--link mdi mdi-delete theme--light"></button>
              <div class="layout flex-row align-center flex-none modal-user-container">
                <div class="flex cols"><button type="button" data-cy="update" class="v-icon notranslate mx-1 ma-2 v-icon--link mdi mdi-pencil theme--light"></button>
                  <div role="dialog" class="v-dialog__container">
                    <!---->
                  </div>
                </div>
              </div>
              <div class="layout flex-row align-center flex-none modal-user-container">
                <div class="flex cols"><button type="button" data-cy="readonly" class="v-icon notranslate mx-1 ma-2 v-icon--link mdi mdi-file-find theme--light"></button>
                  <div role="dialog" class="v-dialog__container">
                    <!---->
                  </div>
                </div>
              </div>
            </td>
          </tr>
          <tr>
            <td data-cy="id"><span class="text-highlight">2</span></td>
            <td data-cy="firstName"><span class="text-highlight">iadh</span></td>
            <td data-cy="lastName"><span class="text-highlight">ben abderrahman</span></td>
            <td data-cy="email"><span class="text-highlight">iadh@email.com</span></td>
            <td data-cy="address"><span class="text-highlight">30 sproles</span></td>
            <td class="actions-container"><button type="button" class="v-icon notranslate table-actions v-icon--link mdi mdi-delete theme--light"></button>
              <div class="layout flex-row align-center flex-none modal-user-container">
                <div class="flex cols"><button type="button" data-cy="update" class="v-icon notranslate mx-1 ma-2 v-icon--link mdi mdi-pencil theme--light"></button>
                  <div role="dialog" class="v-dialog__container">
                    <!---->
                  </div>
                </div>
              </div>
              <div class="layout flex-row align-center flex-none modal-user-container">
                <div class="flex cols"><button type="button" data-cy="readonly" class="v-icon notranslate mx-1 ma-2 v-icon--link mdi mdi-file-find theme--light"></button>
                  <div role="dialog" class="v-dialog__container">
                    <!---->
                  </div>
                </div>
              </div>
            </td>
          </tr>
        </tbody>
      </table>
    </div>
    <div class="v-data-footer">
      <div class="v-data-footer__select">Rows per page:<div class="v-input v-input--hide-details v-input--is-label-active v-input--is-dirty theme--light v-text-field v-select">
          <div class="v-input__control">
            <div role="button" aria-haspopup="listbox" aria-expanded="false" aria-owns="list-42" class="v-input__slot">
              <div class="v-select__slot">
                <div class="v-select__selections">
                  <div class="v-select__selection v-select__selection--comma">10</div><input aria-label="Rows per page:" id="input-42" readonly="readonly" type="text" aria-readonly="false" autocomplete="off">
                </div>
                <div class="v-input__append-inner">
                  <div class="v-input__icon v-input__icon--append"><i aria-hidden="true" class="v-icon notranslate mdi mdi-menu-down theme--light"></i></div>
                </div><input type="hidden" value="10">
              </div>
              <div class="v-menu">
                <!---->
              </div>
            </div>
          </div>
        </div>
      </div>
      <div class="v-data-footer__pagination">1-2 of 2</div>
      <div class="v-data-footer__icons-before"><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Previous page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-left theme--light"></i></span></button></div>
      <div class="v-data-footer__icons-after"><button type="button" disabled="disabled" class="v-btn v-btn--disabled v-btn--icon v-btn--round v-btn--text theme--light v-size--default" aria-label="Next page"><span class="v-btn__content"><i aria-hidden="true" class="v-icon notranslate mdi mdi-chevron-right theme--light"></i></span></button></div>
    </div>
  </div>
</div>
`;
